src/main/java/frc/robot/subsystems/Navigation.java

```diff
@@ -0,0 +1,86 @@
+import edu.wpi.first.math.estimator.DifferentialDrivePoseEstimator;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.DifferentialDriveKinematics;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.subsystems.vision.VisionIO;
+import frc.robot.subsystems.vision.VisionIOInputsAutoLogged;
+import frc.robot.subsystems.vision.VisionIOReal;
+import frc.robot.subsystems.vision.VisionIOSimulation;
+
+public class Navigation extends SubsystemBase {
+    private DifferentialDriveKinematics kinematics;
+    @AutoLogOutput
+    private DifferentialDrivePoseEstimator poseEstimator;
+    @AutoLogOutput
+    private EstimatedRobotPose estimatedVisionPosition;
+    private VisionIO vision;
+    private VisionIOInputsAutoLogged inputs = new VisionIOInputsAutoLogged();
+    private DoubleSupplier leftDistance;
+    private DoubleSupplier rightDistance;
+    private Supplier<Rotation2d> gyro;
+
+    public Navigation (DoubleSupplier leftDistance, DoubleSupplier rightDistance, Supplier<Rotation2d> gyro) {  // REAL ROBOT
+        switch(Constants.BotConstants.botMode) {
+            case REAL:
+                vision = new VisionIOReal();
+                break;
+            case SIM:
+                vision = new VisionIOSimulation();
+                break;
+            case REPLAY:
+                vision = new VisionIOSimulation(); // ?
+                break;
+        }
+        kinematics = new DifferentialDriveKinematics(Constants.DriveConstants.TRACK_WIDTH);
+        poseEstimator = new DifferentialDrivePoseEstimator(
+            kinematics,
+            new Rotation2d(), 
+            0, 
+            0, 
+            new Pose2d(), 
+            VecBuilder.fill(0.05, 0.05, Units.degreesToRadians(5)), 
+            VecBuilder.fill(0.5, 0.5, Units.degreesToRadians(5))
+        );
+
+       // set suppliers
+       this.leftDistance = leftDistance;
+       this.rightDistance = rightDistance;
+       this.gyro = gyro;
+    }
+
+    public void updateOdometry (Rotation2d gyro, double leftDistanceMeters, double rightDistanceMeters) {
+        poseEstimator.update(gyro, leftDistanceMeters, rightDistanceMeters);
+    }
+
+    public void updateNavVision () {
+        if(inputs.estimatedVisionPose != null) {
+            poseEstimator.addVisionMeasurement(inputs.estimatedVisionPose, inputs.timestampSeconds);
+        }
+    }
+
+    @Override
+    public void periodic () {
+        updateOdometry(
+            gyro.get(), 
+            leftDistance.getAsDouble(), 
+            rightDistance.getAsDouble()
+        );
+        vision.updateInputs(inputs);
+        vision.updatePose(poseEstimator.getEstimatedPosition());
+        if(Constants.BotConstants.botMode == Constants.Mode.REAL) {
+            updateNavVision(); // photon lib says sim camera cannot use photon pose estimator
+        }
```

Comment:
I think we should be able to incorporate vision if the botMode is SIM, especially since we are using AdvantageKit. Even if we aren't getting a Photon pose estimated Pose, we should have some sort of simulated pose estimate from either the sim camera, or we could make our own fake pose estimate by taking the robot's true position and adding noise

Currently `updateNavVision` is used to update our DifferentialDrivePoseEstimator given a vision pose estimate, so let's make this possible in sim (at least theoretically, it doesn't have to work yet as long as it compiles fine)

Reply:
I removed the is real conditional and bot code still compiles. I think i just misread docs on the pose estimator. What they are actually saying is in simulation mode the reference pose for vision must come from the simulated drivetrain and not the real DifferentialDrivePoseEstimator class( That is also reflected in their example code )

I updated the code to reflect that logic (calling yet to be made drive sim logic, which i assume will come from DriveIOSimClass using the DifferentialDriveSimulator class)

Reply:
Might still be wrong however. All photon vision says on the matter is "The simulated drivetrain pose must be separate from the drivetrain estimated pose if a pose estimator is utilized."

Reply:
Might still be wrong however. All photon vision says on the matter is "The simulated drivetrain pose must be separate from the drivetrain estimated pose if a pose estimator is utilized."